### PR TITLE
Reduce memory trashing upon class instantiation

### PR DESCRIPTION
**Pull request by Garanas**
As per https://github.com/FAForever/fa/pull/4536 we learned a lot about memory. One of the aspects is that tables resize as you populate `self`. That introduces a lot of garbage, as it continuously runs a similar C snippet to this:

```c
static void resize (lua_State *L, Table *t, int nasize, int nhsize) {
  int i;
  int oldasize = t->sizearray;
  int oldhsize = t->lsizenode;
  Node *nold;
  Node temp[1];
  if (oldhsize)
    nold = t->node;  /* save old hash ... */
  else {  /* old hash is `dummynode' */
    lua_assert(t->node == G(L)->dummynode);
    temp[0] = t->node[0];  /* copy it to `temp' */
    nold = temp;
    setnilvalue(gkey(G(L)->dummynode));  /* restate invariant */
    setnilvalue(gval(G(L)->dummynode));
    lua_assert(G(L)->dummynode->next == NULL);
  }
  if (nasize > oldasize)  /* array part must grow? */
    setarrayvector(L, t, nasize);
  /* create new hash part with appropriate size */
  setnodevector(L, t, nhsize);  
  /* re-insert elements */
  if (nasize < oldasize) {  /* array part must shrink? */
    t->sizearray = nasize;
    /* re-insert elements from vanishing slice */
    for (i=nasize; i<oldasize; i++) {
      if (!ttisnil(&t->array[i]))
        setobjt2t(luaH_setnum(L, t, i+1), &t->array[i]);
    }
    /* shrink array */
    luaM_reallocvector(L, t->array, oldasize, nasize, TObject);
  }
  /* re-insert elements in hash part */
  for (i = twoto(oldhsize) - 1; i >= 0; i--) {
    Node *old = nold+i;
    if (!ttisnil(gval(old)))
      setobjt2t(luaH_set(L, t, gkey(old)), gval(old));
  }
  if (oldhsize)
    luaM_freearray(L, nold, twoto(oldhsize), Node);  /* free old array */
}
```

Long story short - we learned we can pre-allocate either the hash or table part with entries. This pull request attempts to pre-allocate the projectile table so that it has 16 available entries from the start, instead of having to resize multiple times to get to that point. There are up to four resizes (0 to 40 bytes, 40 to 80, 80 to 160 and 160 to 320), each time a projectile is created.

But it doesn't work yet 😄 !

### CODE REVIEW

**Comment by Garanas** [comment]
What I don't get is why the [Projectile factory](https://github.com/FAForever/fa/pull/4539/files#diff-674904797c709585a0513c6a812e748d8af279fdf8ba59f2bc42ea56112ff5e8R484-R492) is never used, even though we set the correct meta table.

**Comment by Garanas** [comment]
@KionX it appears that projectiles, units and weapons all use the metatable of `IEffect` to instantiate themselves. Could you look into the assembly, and see if it would be do-able to make them use their own meta table, as set during the global init file?

That way we can set the amount to pre-allocate to be different for each of them, without having to introduce a hacky solution in Lua.

**Comment by Garanas** [comment]
On the current release branch:

https://user-images.githubusercontent.com/15778155/209958968-7dcabb19-209b-4274-8407-18da18fd4722.mp4

Same setup, but this branch:

https://user-images.githubusercontent.com/15778155/209959825-95ba88aa-a8fc-4684-8391-b092b795a3a8.mp4

File: lua/maui/control.lua [hunk 1]
```diff
@@ -1,4 +1,4 @@
--- Class methods
+-- ClassUI methods
```
Comment (4z0t) [inline lua/maui/control.lua:1]: Some comments were changed too :)

File: lua/ui/lobby/data/watchedvalue/watchedvaluearray.lua [hunk 2]
```diff
@@ -3,7 +3,7 @@ local WatchedValueTable = import("/lua/ui/lobby/data/watchedvalue/watchedvalueta
 --- A WatchedValueTable with integral keys. Note that `false` is indistinguishable from `nil` in this
 -- structure, because Lua.
 ---@class WatchedValueArray : WatchedValueTable
-WatchedValueArray = Class(WatchedValueTable) {
+WatchedValueArray = ClassUI(WatchedValueTable) {
```
Comment (4z0t) [inline lua/ui/lobby/data/watchedvalue/watchedvaluearray.lua:6]: This isnt UI class.

File: lua/ui/lobby/data/playerdata.lua [hunk 3]
```diff
@@ -52,7 +52,7 @@ local DEFAULT_MAPPING = {
 
 -- Represents player data using the magic of lazy variables.
 ---@class WatchedPlayerData : PlayerData, WatchedValueTable
-PlayerData = Class(WatchedValueTable) {
+PlayerData = ClassUI(WatchedValueTable) {
```
Comment (4z0t) [inline lua/ui/lobby/data/playerdata.lua:55]: This isnt UI class.

File: lua/maui/layouthelpers.lua [hunk 4]
```diff
@@ -2194,7 +2194,7 @@ local LayouterAttributeSelection = ClassSimple {
 
 ---@class Layouter : LayouterAttributeControl, LayouterAttributeDropShadow, LayouterAttributeEditor, LayouterAttributeTexture, LayouterAttributeSelection
 ---@field layoutControl Control
-Layouter = Class(LayouterAttributeControl, LayouterAttributeDropShadow, LayouterAttributeEditor, LayouterAttributeTexture, LayouterAttributeSelection) {
+Layouter = ClassUI(LayouterAttributeControl, LayouterAttributeDropShadow, LayouterAttributeEditor, LayouterAttributeTexture, LayouterAttributeSelection) {
```
Comment (4z0t) [inline lua/maui/layouthelpers.lua:2197]: This isnt UI class.

File: lua/maui/layouthelpers.lua [hunk 5]
```diff
@@ -2120,7 +2120,7 @@ local LayouterAttributeColor = ClassSimple {
 ------------------------------
 
 ---@class LayouterAttributeTexture : LayouterAttributeColor
-local LayouterAttributeTexture = Class(LayouterAttributeColor) {
+local LayouterAttributeTexture = ClassUI(LayouterAttributeColor) {
```
Comment (4z0t) [inline lua/maui/layouthelpers.lua:2123]: This isnt UI class.

File: lua/maui/layouthelpers.lua [hunk 6]
```diff
@@ -2073,7 +2073,7 @@ local LayouterAttributeFont = ClassSimple {
 ------------------------------
 
 ---@class LayouterAttributeEditor : LayouterAttributeFont
-local LayouterAttributeEditor = Class(LayouterAttributeFont) {
+local LayouterAttributeEditor = ClassUI(LayouterAttributeFont) {
```
Comment (4z0t) [inline lua/maui/layouthelpers.lua:2076]: This isnt UI class.

**Review by 4z0t — APPROVED** [review]

**Comment by Garanas** [comment]
I'll continue tomorrow with the other classes, such as units and weapons

**Pull request merged**